src/tree/updater_gpu_hist.cu

```diff
@@ -1328,23 +1329,47 @@ inline void DeviceShard<GradientSumT>::InitCompressedData(
 
 template <typename GradientSumT>
 inline void DeviceShard<GradientSumT>::CreateHistIndices(
-    const SparsePage& row_batch, size_t row_stride, int null_gidx_value) {
+    const SparsePage &row_batch,
+    const common::HistCutMatrix &hmat,
+    const RowStateOnDevice &device_row_state,
+    int rows_per_batch) {
+  // Has any been allocated for me in this batch?
+  if (!device_row_state.batch_n_rows) return;
+
+  size_t row_stride = this->ellpack_matrix.row_stride;
+
+  const auto &offset_vec = row_batch.offset.ConstHostVector();
+  /*! \brief row offset in SparsePage (the input data). */
+  thrust::device_vector<size_t> row_ptrs(device_row_state.batch_n_rows+1);
+  thrust::copy(
+    offset_vec.data() + device_row_state.row_offset,
+    offset_vec.data() + device_row_state.row_offset + device_row_state.batch_n_rows + 1,
+    row_ptrs.begin());
+
   int num_symbols = n_bins + 1;
   // bin and compress entries in batches of rows
-  size_t gpu_batch_nrows =
+  size_t gpu_batch_nrows = 0;
+  if (rows_per_batch == 0) {
```

Comment:
Let's have some comments to explain these three conditions.

Reply:
will do...

Reply:
The reason we have this for quantiles is that the batch size affects the exact quantile values that are found.

There is no such problem here, i.e. the result does not depend on the batch size. Therefore, there's no need for a complicated selection of `rows_per_batch`, the automatically selected value will do, as `row_stride` is already computed based on all batches and shards.

Reply:
done